internal/service/mock/upcloud_client.go

```diff
@@ -0,0 +1,108 @@
+package mock
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/UpCloudLtd/upcloud-go-api/v6/upcloud"
+	"github.com/UpCloudLtd/upcloud-go-api/v6/upcloud/request"
+	upsvc "github.com/UpCloudLtd/upcloud-go-api/v6/upcloud/service"
+)
+
+type UpCloudClient struct {
+	upsvc.Storage
+
+	servers sync.Map
+}
+
+func (u *UpCloudClient) StoreServer(s *upcloud.ServerDetails) {
+	u.servers.LoadOrStore(s.UUID, s)
+}
+
+func (u *UpCloudClient) getServer(id string) *upcloud.ServerDetails {
+	if s, ok := u.servers.Load(id); ok {
+		return s.(*upcloud.ServerDetails)
+	}
+	return nil
+}
+
+func (u *UpCloudClient) WaitForServerState(ctx context.Context, r *request.WaitForServerStateRequest) (*upcloud.ServerDetails, error) {
+	s, _ := u.GetServerDetails(ctx, &request.GetServerDetailsRequest{
+		UUID: r.UUID,
+	})
+	return s, nil
+}
+
+func (u *UpCloudClient) GetServers(ctx context.Context) (*upcloud.Servers, error) {
+	s := []upcloud.Server{}
+	u.servers.Range(func(key, value any) bool {
+		if d, ok := value.(*upcloud.ServerDetails); ok {
+			s = append(s, d.Server)
+		}
+		return true
+	})
+	return &upcloud.Servers{Servers: s}, nil
+}
+
+func (u *UpCloudClient) GetServerDetails(ctx context.Context, r *request.GetServerDetailsRequest) (*upcloud.ServerDetails, error) {
+	if s := u.getServer(r.UUID); s != nil {
+		return s, nil
+	}
+	return nil, fmt.Errorf("server '%s' not found", r.UUID)
+}
+
+func (u *UpCloudClient) AttachStorage(ctx context.Context, r *request.AttachStorageRequest) (*upcloud.ServerDetails, error) {
+	server := u.getServer(r.ServerUUID)
+	if server == nil {
+		return server, errors.New("server not found")
+	}
+	if server.State != upcloud.ServerStateStarted {
+		return nil, fmt.Errorf("server %s state is %s", r.ServerUUID, server.State)
+	}
+	server.State = upcloud.ServerStateMaintenance
+	u.StoreServer(server)
+	time.Sleep(time.Duration(rand.Intn(200)+100) * time.Millisecond) //nolint:gosec // using weak random number doesn't affect the result.
```

Comment:
Is this needed if we have that nodeSync mutex anyway?

Reply:
@ajmyyra thanks for the review :+1:  
It's not needed per se, I just wanted to simulate some latency in response times that we might have from the backend. In the `TestUpCloudService_AttachDetachStorage_Concurrency` test case we run attach operation against two different server UUIDs which initializes two different mutex locks. This timeout tries to make sure that our function can handle  concurrent requests also with multiple locks. Does that make any sense or was I just overthinking this ? :)

Reply:
Sounds good, no problems here. :)